### ISSUE TITLE
Integrate sign and verify to PBFT

### ISSUE THREAD

**Issue by kchristidis** [hyperledger-archives/fabric#273]
We need to make sure that every message that needs to be signed or verified is properly signed or verified.

The #241 PR may be a good starting point.

**Comment by kchristidis** [comment]
@tuand27613 this will probably be moved to you, depending on the progress of the other issues. Just adding this as a reminder here.

**Comment by corecode** [comment]
It is unclear to me what this issue is about exactly.  We have several interfaces that require some form of crypto:
- [ ] view-change public-key signatures
- [ ] checkpoint public-key signatures
- [ ] later: pre-execution (weed out non deterministic transactions) signatures
- [ ] TLS certificate checks to establish identity and message authenticity (connection layer)

Should we use this issue to collect progress on all these fronts?

**Comment by kchristidis** [comment]
The original intention of this issue was to just link to the methods that @adecaro had created for signing/verifying in the #241 PR, as a placeholder of sorts until the actual crypto kicks in.

We can however use this issue to collect progress on the items you list above, sure.

**Comment by corecode** [comment]
Need example code from @adecaro how to sign and verify messages from other replicas.  Specifically, how do we get the `vkID` for `Verify`, and how do we verify that the `vkID` actually belongs to a validator.

**Comment by corecode** [comment]
Also waiting on backend support for `TransactionPreValidation`: validator needs to be initialized and be accessible somewhere in the stack.

**Comment by tuand27613** [comment]
won't be ready week ending 1/15.
Got twisted around between obc verify versus obc_message verify.
Trying to get it somewhat tested by friday.

**Comment by tuand27613** [comment]
The majority of the work is in [consensus/obcpbft/sign.go](https://github.com/openblockchain/obc-peer/blob/master/openchain/consensus/obcpbft/sign.go)  where all the needed functions have been stubbed out. We need to now call the crypto functions sign() and verify() while getting the necessary parameters from the stack, e.g. correct peer ID and certs. Also, we'll want to handle the encrypt/not encrypt flag.

We also want to move the sign/verify functions out of obcpbft since we want them to be available to any consensus plugin. Current thought is to put into CPI/helper. I'll update when I decide for sure.

Note that not all messages are signed. As @corecode mentioned, this is for performance reasons. Actual   messages types that need sign/verify are defined in [obcpbft/consensus/obcpbft/messages.proto](https://github.com/openblockchain/obc-peer/blob/master/openchain/consensus/obcpbft/messages.proto) with a signature field.  When sign/verify functions are called is controlled via the code in sign.go and the signable interface.

**Comment by tuand27613** [comment]
Note that there are 2 different verify functions in consensus package, the sign/verify crypto one and the one that does transaction prevalidation ( see e.g. verify() in obc-classic RecvMsg() function ). The pre validation code work should be treated as a separate issue ?

**Comment by kchristidis** [comment]
@tuand27613 Let's split out the transaction pre-validation work into a separate issue please, yes.

**Comment by tuand27613** [comment]
Needed crypto hash to be included in PeerEndPoint so that we have access to the hash when calling verify(). @jeffgarratt will update the Peer code under: https://github.com/openblockchain/obc-peer/issues/514

**Comment by kchristidis** [comment]
@tuand27613 has completed and tested this. There is a minor ssue with the comparison of `PeerID`s that I'll investigate today before we can close this. PR coming before EOD.

**Comment by kchristidis** [comment]
`PeerID` issue resolved but there's an [outstanding `TODO`](https://github.com/tuand27613/obc-peer/commit/264d4b08ccd567db741e90c33ff6e43f91db16dd#commitcomment-15681955) that @tuand27613 is looking into.

**Comment by kchristidis** [comment]
Fixed in PR #565.
